sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -519,6 +519,7 @@ case class ScalaAggregator[IN, BUF, OUT](
   def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ScalaAggregator[IN, BUF, OUT] =
     copy(inputAggBufferOffset = newInputAggBufferOffset)
 
+  @transient
   private[this] lazy val inputProjection = UnsafeProjection.create(children)
```

Comment:
there are some other lazy vals in this class, why only this field is an issue?

Reply:
thanks for looking. lazy vals are not a problem in general if they are serializable, the problem is that this inputProjection field holds an instance of GeneratedClass$SpecificUnsafeProjection, which is not serializable and fails the serialization of AggregatingAccumulator used by observe (AggregatingAccumulator -> typedImperatives -> ScalaAggregator -> inputProjection).

Reply:
Does this mean this lazy val gets accessed at the driver side, then we serialize and send the UDAF to the executor? Do you know the whole workflow?

Reply:
As far as i get it it's not being accessed in the driver.
This field is accessed and initialized in the executor, as part of ScalaAggregator#update, and then causes the failure when trying to return a serialized TaskResult to the driver (which holds the mentioned accumulator).

Reply:
I see, seems we can mark all lazy vals in this class as `transient` they can be recomputed at the driver side?

Reply:
Yes seems so, annotated all of them now